app/models/html_attachment.rb

```diff
@@ -68,9 +68,19 @@ def url(options = {})
 
     type = attachable.path_name
     path_or_url = full_url ? "url" : "path"
-    path_helper = "#{type}_html_attachment_#{path_or_url}"
 
-    Whitehall.url_maker.public_send(path_helper, attachable.slug, self, options)
+    # This depends on the rails url helpers to construct a url based on config/routes.rb:
+    # composed of document type, slug for the parent document, and identifier for the attachment;
+    # for non-english attachments the identifier is the content_id, for english
+    # attachments it is self i.e. the slug
+    if sluggable_locale? || attachable.updated_at < Time.zone.local(2021, 11, 29)
+      identifier = self
+      path_helper = "#{type}_html_attachment_#{path_or_url}"
+    else
+      identifier = content_id
+      path_helper = "#{type}_html_attachment_by_content_id_#{path_or_url}"
+    end
+    Whitehall.url_maker.public_send(path_helper, attachable.slug, identifier, options)
```

Comment:
My final comment is a bit of a spanner in the works, but hopefully could lead to a much simpler solution. Do we need to make the distinction between the different path helpers? Could we not do something like this instead?

routes.rb:

```ruby
get "/consultations/:consultation_id/public-feedback/:edition_or_content_id" => "_#_", as: "consultation_public_feedback_html_attachment"
```

html_attachment.rb:

```ruby
identifier = sluggable_locale? ? self : content_id
Whitehall.url_maker.public_send("consultation_public_feedback_html_attachment", attachable.slug, identifier, options)
```

And then the missing piece of the puzzle in the controller would be something like:

```ruby
edition = Edition.find_by(id: edition_or_content_id) || Edition.find_by(content_id: edition_or_content_id) # and maybe even || Edition.find_by(slug: edition_or_content_id)
```

That way, we wouldn't need to have an out-of-hours deploy, a timing based cut-off point, or maintain two different route styles in routes.rb. We'd be encapsulating the different ID types in the controller and handling it there. Does that make sense?

Reply:
This certainly seems preferable yeah, I just had trouble making sense of the routes and how they bound to controller methods.

It didn't occur to me that this would allow us to avoid the timing-based cut-off, though: I thought that the point was that editions published before the cutover date should continue to have old-style URLs.

If we don't mind the new-style URLs appearing immediately, then I think we can just drop the commit adding that distinction. Having both the old-style route and the new-style route in place would still achieve that. (To be clear we should _also_ make this change to avoid having multiple routes — I just think the two things can be decoupled.)

Reply:
I think with the pseudocode snippets above, the old-style URLs should still work.

So for example, https://www.gov.uk/government/publications/universal-credit-and-rented-housing--2/5654420 would continue to resolve. But with this change, the same HTML attachment would also be accessible at https://www.gov.uk/government/publications/universal-credit-and-rented-housing--2/2c9002ee-19fd-4202-9720-b70aa8042126; the controller would just attempt to find by Edition ID first, fail to do that, and then try to find by Content ID.

**Subsequent publishes** of the parent edition would cause a switch to the new-style URL, because that is when the `consultation_public_feedback_html_attachment` helper would get used. But that's intentional behaviour (and something we'll be relying on in the 2nd card when we batch-republish all applicable documents). Does that make sense?

I should caveat this with "this code is untested" - I _think_ the approach would work, but we'd want to spike it first. If we took it in this direction, I'd suggest probably attempting it in a separate branch & PR.

Reply:
Presumably we could leave the double routes in place and then figure out if we can tidy that up as part of reverting the unnecessary parts after the migration?

On the other hand, if I can get it working now, then there'd be no need to revert it anyway. So I think I'll try to figure out if I can make sense of how the routes call the controllers (`_#_` is unhelpfully difficult to google for).

Reply:
From Andrew White, via Slack:

> If those route actually get hit they’ll look for a `Controller` class with an action `_` - that implies they’re being used to generate custom url helpers. There’s a relatively new method for defining things like that called `direct` - you can probably switch to that for these helpers https://api.rubyonrails.org/classes/ActionDispatch/Routing/Mapper/CustomUrls.html#method-i-direct

Reply:
Ah, I think what I'd somehow missed/forgotten is that these HTML attachments are rendered by government-frontend. 

So, given an [example non-English HTML attachment](https://www.gov.uk/government/publications/universal-credit-and-rented-housing--2/universal-credit-and-rented-housing-guide-for-landlords), if we look at its [content item](https://www.gov.uk/api/content/government/publications/universal-credit-and-rented-housing--2/universal-credit-and-rented-housing-guide-for-landlords), we can see `"rendering_app": "government-frontend"`.

As far as I can gather, government-frontend passes most requests to `ContentItemsController` with a [catch-all route](https://github.com/alphagov/government-frontend/blob/b986beb486d8de29703449d5b2f0d70f12c70020/config/routes.rb#L21-L24). This [fetches the content item via the request path](https://github.com/alphagov/government-frontend/blob/7e3397b98d091259c286cc9a52156e0338518f0e/app/controllers/content_items_controller.rb#L90-L101), which basically does a `Document.find_by(path: THE_PATH)` on Content Store. So we could publish _any_ route, so long as we tell Publishing API that the rendering app is Government Frontend.

I've taken your first commit, removed any routes.rb changes (as they're not needed), and raised #6377. I've tested this on Integration and can confirm that the Welsh HTML attachment's URL changes to content-ID based URL, a redirect is applied from the old attachment to the new, and English attachments are unaffected.